presto-main/src/main/java/com/facebook/presto/server/remotetask/HttpRemoteTaskFactory.java

```diff
@@ -144,4 +154,40 @@ public RemoteTask createRemoteTask(Session session,
                 partitionedSplitCountTracker,
                 stats);
     }
+
+    @Override
+    public void destroyExchangeSources(Set<URI> locationsToDestroy)
+    {
+        for (URI location : locationsToDestroy) {
+            Request request = prepareDelete()
+                    .setUri(location)
+                    .build();
+            executeDestroyExchangeSourceRequest(createCleanupBackoff(), request);
+        }
+    }
+
+    private void executeDestroyExchangeSourceRequest(Backoff cleanupBackoff, Request request)
+    {
+        addExceptionCallback(httpClient.executeAsync(request, createFullJsonResponseHandler(taskInfoCodec)), throwable -> {
+            if (throwable instanceof RejectedExecutionException && httpClient.isClosed()) {
+                logError(throwable, "Unable to destroy exchange source at %s. HTTP client is closed.", request.getUri());
+                return;
+            }
+
+            // record failure
+            if (cleanupBackoff.failure()) {
+                logError(throwable, "Unable to destroy exchange source at %s. Back off depleted.", request.getUri());
```

Comment:
This is problematic. This operation is different from task abort operations.

When we try to abort a task, we'll just give up, and let the task on the worker expire. However, here, the task is still healthy from the coordinator's point of view. As a result, giving up would cause the query to dead lock and last forever.

I propose failing the task here.